community/common/src/main/java/org/neo4j/kernel/lifecycle/SafeLifecycle.java

```diff
@@ -0,0 +1,180 @@
+ *  -: Error    (will throw IllegalStateException)
+ *
+ *  P: PRE      ("pre init")
+ *  I: IDLE     ("initialized" or "stopped")
+ *  R: RUN      ("started")
+ *  H: HALT     ("shutdown")
+ * </pre>
+ * A successful operation is an operation not throwing an exception.
+ * <p>
+ * End states on a successful operation.
+ * <pre>
+ *  ---------------------------------------------------
+ * | FROM \ op |  init()  start()  stop()  shutdown()  |
+ *  ---------------------------------------------------
+ * | PRE       |   IDLE      -       -       HALT(*)   |
+ * | IDLE      |    -       RUN    IDLE(*)    HALT     |
+ * | RUN       |    -        -      IDLE       -       |
+ * | HALT      |    -        -       -         -       |
+ *  ---------------------------------------------------
+ * </pre>
+ * End states on a failed operation.
+ * <pre>
+ *  ---------------------------------------------------
+ * | FROM \ op |  init()  start()  stop()  shutdown()  |
+ *  ---------------------------------------------------
+ * | PRE       |   PRE       -       -       HALT(*)   |
+ * | IDLE      |    -       IDLE   IDLE(*)    HALT     |
+ * | RUN       |    -        -      IDLE       -       |
+ * | HALT      |    -        -       -         -       |
+ *  ---------------------------------------------------
+ * </pre>
+ * A few notes:
+ * <ul>
+ * <li>will not invoke stop0() if start0() wasn't successful</li>
+ * <li>will not invoke shutdown0() if init0() wasn't successful</li>
+ * </ul>
+ * The expectation with regards to error handling and cleanup is that
+ * an unclean start() is cleaned up by the start0() method and thus
+ * the component is left in IDLE. The same goes for issues happening
+ * during init0(), which leaves the component in PRE.
+ * <p>
+ * Because of the way that {@link LifeSupport} operates today, this
+ * class will ignore stop() calls made while in IDLE. Similarly, calls
+ * to shutdown() will be ignored while in PRE. This allows this class
+ * to be managed by a {@link LifeSupport} without throwing
+ * {@link IllegalStateException} on those state transitions, which
+ * otherwise would have been disallowed and handled in the same way
+ * as other illegal state transitions.
+ * <p>
+ * This adapter will not allow a shutdown lifecycle to be reinitialized
+ * and started again.
+ */
+public abstract class SafeLifecycle implements Lifecycle
+{
+    private State state = PRE;
+
+    /**
+     * @param expected The expected from state.
+     * @param to The to state.
+     * @param op The state transition operation.
+     * @param force Causes the state to be updated regardless of if the operation throws.
+     * @throws Throwable Issues.
+     */
+    private void transition( State expected, State to, Operation op, boolean force ) throws Throwable
+    {
+        if ( state != expected )
+        {
+            throw new IllegalStateException( String.format( "Expected %s but was %s", expected, state ) );
+        }
+
+        if ( force )
```

Comment:
Would it be worth logging a warning on a force?

Reply:
This is an expected path so logging a warning would be confusing. Perhaps the parameter should be renamed. The intention is that the state transition update happens regardless of whether the operation is successful. For example when you call stop(), you should not stay in RUN even if stop() throws midway. The same for shutdown(). This is different than for start() which will not cause a transition on a failure.

Reply:
Ah I see it now - makes sense.